### PR TITLE
DJORNL: add in a parser check for missing files

### PR DESCRIPTION
**Pull request by ialarmedalien**
Ping me when you get to this PR as it needs to be done after #50, and I will have to rebase and push again.

Add in a parser check for missing headers.

- [x] I updated the README.md docs to reflect this change. - N/A
- [x] This is not a breaking API change OR

### CODE REVIEW

File: importers/djornl/parser.py [hunk 1]
```diff
@@ -144,9 +155,46 @@ def parser_gen(self, file):
                 msg = f"expected {expected_col_count} cols, found {col_count}"
                 yield(line_no, None, f"{file['path']} line {line_no}: {msg}")
 
+    def check_headers(self, headers, validator=None):
+        """
+        Ensure that the file headers contain required columns for the data type. Checks the schema
+        in the validator to ensure that all required fields are present in the headers.
```
Comment (ialarmedalien) [inline importers/djornl/parser.py:161]: prevents files from throwing an error every line if the headers are missing a required field

File: importers/djornl/parser.py [hunk 2]
```diff
@@ -189,6 +237,12 @@ def process_file(self, file, remap_fn, store_fn, err_list, validator=None):
             err_list.append(f"{file['path']}: no header line found")
             return
 
+        missing_headers = self.check_headers(cols, validator)
```
Comment (ialarmedalien) [inline importers/djornl/parser.py:240]: This is in the `process_file` function; after ensuring that the file does actually have content, the first valid (non-comment) line of the file is assumed to be the headers, and is checked against the spec to ensure that all required fields are present.

File: importers/djornl/parser.py [hunk 3]
```diff
@@ -210,8 +264,15 @@ def process_file(self, file, remap_fn, store_fn, err_list, validator=None):
...
-            # transform it using the remap_functions
-            datum = self.remap_object(row_object, remap_fn)
+            try:
+                # transform it using the remap_functions
+                datum = self.remap_object(row_object, remap_fn)
+            except Exception as err:
+                err_type = type(err)
+                err_list.append(
+                    f"{file['path']} line {line_no}: error remapping data: {err_type} {err}"
+                )
```
Comment (ialarmedalien) [inline importers/djornl/parser.py:274]: prevents mysterious uninformative errors during the remapping of parsed data ==> storable data

File: importers/djornl/parser.py [hunk 4]
```diff
@@ -239,22 +300,24 @@ def load_edges(self):
...
+            'score': lambda row: float(row['score']),
+            'edge_type': None,
         }
 
         # store edge data, checking for potential duplicates
         def store_edges(datum):
             # there should only be one value for each node<->node edge of a given type,
             # so use these values as an index key
-            edge_key = "__".join([datum['node1'], datum['node2'], datum['edge_type']])
+            edge_key = "__".join([*sorted([datum['node1'], datum['node2']]), datum['edge_type']])
```
Comment (ialarmedalien) [inline importers/djornl/parser.py:320]: Ensures that edge A-->B of type Z is recognised as being the same as edge B-->A of type Z as the edge direction is meaningless in this context.

File: importers/test/test_djornl_parser.py [hunk 5]
```diff
@@ -72,6 +72,15 @@ def test_load_invalid_file(self):
         with self.assertRaisesRegex(RuntimeError, err_str):
             self.init_parser_with_path(RES_ROOT_DATA_PATH)
 
+    def test_load_missing_files(self):
```
Comment (ialarmedalien) [inline importers/test/test_djornl_parser.py:75]: just moving this test up in the file

**Comment by jayrbolton** [comment]
:+1:

**Comment by jayrbolton** [comment]
My bad, I assumed this would go into develop, since it was listed as the first PR.

**Pull request merged**